File: android/app/src/main/java/com/mail/client/data/local/MessageDao.kt
package com.mail.client.data.local

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction
import kotlinx.coroutines.flow.Flow

@Dao
interface MessageDao {

    @Query("SELECT * FROM messages WHERE threadId = :threadId ORDER BY date ASC")
    fun observeForThread(threadId: String): Flow<List<MessageEntity>>

    @Query("SELECT * FROM messages WHERE threadId = :threadId ORDER BY date ASC")
    suspend fun getForThread(threadId: String): List<MessageEntity>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(messages: List<MessageEntity>)

    @Query("DELETE FROM messages WHERE threadId = :threadId")
    suspend fun deleteForThread(threadId: String)

    /** Atomically replaces all messages for a thread — Flow observers never see an empty state. */
    @Transaction
    suspend fun replaceForThread(threadId: String, messages: List<MessageEntity>) {
        deleteForThread(threadId)
        insertAll(messages)
    }
}

File: flutter/lib/data/local/message_dao.dart
part of 'mail_database.dart';

@DriftAccessor(tables: [Messages])
class MessageDao extends DatabaseAccessor<MailDatabase> with _$MessageDaoMixin {
  MessageDao(super.db);

  /// Live stream of messages for a thread, oldest first.
  Stream<List<Message>> watchForThread(String threadId) =>
      (select(messages)
            ..where((m) => m.threadId.equals(threadId))
            ..orderBy([(m) => OrderingTerm.asc(m.date)]))
          .watch();

  Future<List<Message>> getForThread(String threadId) =>
      (select(messages)
            ..where((m) => m.threadId.equals(threadId))
            ..orderBy([(m) => OrderingTerm.asc(m.date)]))
          .get();

  Future<void> insertAll(List<MessagesCompanion> rows) =>
      batch((b) => b.insertAllOnConflictUpdate(messages, rows));

  Future<void> deleteForThread(String threadId) =>
      (delete(messages)..where((m) => m.threadId.equals(threadId))).go();

  /// Atomically replaces all messages for a thread.
  /// Wrapping in a transaction means Flow observers never see an intermediate
  /// empty state — mirrors MessageDao.replaceForThread() in Room.
  Future<void> replaceForThread(
    String threadId,
    List<MessagesCompanion> rows,
  ) =>
      transaction(() async {
        await deleteForThread(threadId);
        await insertAll(rows);
      });
}

// ── Extension helpers (mirrors MessageEntity.labelList()) ─────────────────────

extension MessageX on Message {
  List<String> get labelList =>
      labelIds.isEmpty ? [] : labelIds.split(',');
}
